drivers/hwspinlock/sqn_hwspinlock.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2023 Sequans Communications
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT sqn_hwspinlock
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/sys_io.h>
+#include <zephyr/drivers/hwspinlock.h>
+
+#include <zephyr/sys/printk.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sqn_hwspinlock);
+
+struct sqn_hwspinlock_data {
+	DEVICE_MMIO_RAM;
+	uint32_t num_locks;
+};
+
+struct sqn_hwspinlock_config {
+	DEVICE_MMIO_ROM;
+	const struct device *hwspinlocks;
+};
+
+static inline mem_addr_t get_lock_addr(const struct device *dev, uint32_t id)
+{
+	return (mem_addr_t)(DEVICE_MMIO_GET(dev) + id * sizeof(uint32_t));
+}
+
+/*
+ * To define CPU id, we use the affinity2 and affinity1
+ * fields of the MIPID.
+ */
+static uint8_t mpidr_to_cpuid(uint64_t mpidr_val)
+{
+	uint8_t cpuid = ((mpidr_val >> 8) & 0x0F) | ((mpidr_val >> 12) & 0xF0);
+	return ++cpuid;
+}
+
+static int sqn_hwspinlock_trylock(const struct device *dev, uint32_t id)
+{
+	struct sqn_hwspinlock_data *data = dev->data;
+	uint8_t cpuid;
+
+	if(id > data->num_locks)
+		return -EINVAL;
+
+	cpuid = mpidr_to_cpuid(read_mpidr_el1());
+	if (sys_read8(get_lock_addr(dev, id)) == cpuid)
+		return 0;
+
+	sys_write8(cpuid, get_lock_addr(dev, id));
+	if (sys_read8(get_lock_addr(dev, id)) == cpuid)
+		return 0;
+	else
+		return -EBUSY;
+}
+
+static void sqn_hwspinlock_lock(const struct device *dev, uint32_t id)
+{
+	struct sqn_hwspinlock_data *data = dev->data;
+	uint8_t cpuid;
+
+	if (id > data->num_locks)
+		return;
+
+	cpuid = mpidr_to_cpuid(read_mpidr_el1());
+	if(sys_read8(get_lock_addr(dev, id)) == 0)
+		sys_write8(cpuid, get_lock_addr(dev, id));
+
+	while(sys_read8(get_lock_addr(dev, id)) != cpuid) {
+		k_sleep(K_NSEC(CONFIG_SQN_HWSPINLOCK_RELAX_TIME));
+		sys_write8(cpuid, get_lock_addr(dev, id));
+	}
+}
+
+static void sqn_hwspinlock_unlock(const struct device *dev, uint32_t id)
+{
+	struct sqn_hwspinlock_data *data = dev->data;
+	uint8_t cpuid;
+
+	if (id > data->num_locks)
+		return;
+
+	cpuid = mpidr_to_cpuid(read_mpidr_el1());
+	sys_write8(cpuid, get_lock_addr(dev, id));
```

Comment:
You write the same value back to the register to **un**lock?  Not a zero or some other token?  How does the hardware distinguish this from the locking protocol, which does the same thing?

Reply:
Please check comment above about how our hwspinlock works.

If the register value is equal to the cpuid and we write the cpuid, then the register value will be 0.
So to unlock the hwspinlock, we write cpuid.

Reply:
I added a comment in source code to explain the hwspinlock unlocking.